File: intellij-lsp/src/com/github/gtache/requests/ReformatHandler.scala
package com.github.gtache.requests

import com.github.gtache.editor.EditorEventManager
import com.github.gtache.PluginMain
import com.github.gtache.utils.Utils
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.{FileEditorManager, OpenFileDescriptor}
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ProjectFileIndex
import com.intellij.openapi.vfs.VirtualFile

object ReformatHandler {

  def reformatAllFiles(project: Project): Boolean = {
    var allFilesSupported = true
    ProjectFileIndex.getInstance(project).iterateContent((fileOrDir: VirtualFile) => {
      if (fileOrDir.isDirectory) {
        true
      } else {
        if (PluginMain.isExtensionSupported(fileOrDir.getExtension)) {
          reformatFile(fileOrDir, project)
          true
        } else {
          allFilesSupported = false
          true
        }
      }
    })
    allFilesSupported
  }

  def reformatFile(file: VirtualFile, project: Project): Unit = {
    val uri = Utils.VFSToURIString(file)
    EditorEventManager.forUri(uri) match {
      case Some(manager) =>
        manager.reformat()
      case None =>
        val fileEditorManager = FileEditorManager.getInstance(project)
        val descriptor = new OpenFileDescriptor(project, file)
        ApplicationManager.getApplication.invokeLater(() => ApplicationManager.getApplication.runWriteAction(new Runnable {
          override def run(): Unit = fileEditorManager.openTextEditor(descriptor, false)
        }))
        while (EditorEventManager.forUri(uri).isEmpty) {}
        EditorEventManager.forUri(uri).get.reformat()
        ApplicationManager.getApplication.invokeLater(() => ApplicationManager.getApplication.runWriteAction(new Runnable {
          override def run(): Unit = fileEditorManager.closeFile(file)
        }))
    }
  }

  def reformatFile(editor: Editor): Unit =
    EditorEventManager.forEditor(editor).foreach(manager => manager.reformat())


  def reformatSelection(editor: Editor): Unit = {
    EditorEventManager.forEditor(editor).foreach(manager => manager.reformatSelection())
  }

}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/requests/ReformatHandler.kt
package com.github.gtache.lsp.requests

import com.github.gtache.lsp.LSPProjectService
import com.github.gtache.lsp.editor.EditorApplicationService
import com.github.gtache.lsp.editor.EditorProjectService
import com.github.gtache.lsp.utils.ApplicationUtils
import com.github.gtache.lsp.utils.FileUtils
import com.intellij.openapi.components.service
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.OpenFileDescriptor
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ProjectFileIndex
import com.intellij.openapi.vfs.VirtualFile

/**
 * Object handling reformat events
 */
object ReformatHandler {

    /**
     * Unused
     * Reformats all the files in the project
     *
     * @param project The project
     * @return True if all the files were supported by the language servers, false otherwise
     */
    fun reformatAllFiles(project: Project): Boolean {
        var allFilesSupported = true
        ProjectFileIndex.getInstance(project).iterateContent { fileOrDir ->
            if (fileOrDir.isDirectory) {
                true
            } else {
                if (project.service<LSPProjectService>().isExtensionSupported(fileOrDir.extension)) {
                    reformatFile(fileOrDir, project)
                    true
                } else {
                    allFilesSupported = false
                    true
                }
            }
        }
        return allFilesSupported
    }

    /**
     * Reformat a file given a VirtualFile and a Project
     *
     * @param file    The file
     * @param project The project
     */
    fun reformatFile(file: VirtualFile, project: Project): Unit {
        if (project.service<LSPProjectService>().isExtensionSupported(file.extension)) {
            val uri = FileUtils.VFSToURI(file)
            if (uri != null) {
                val manager = project.service<EditorProjectService>().forUri(uri)
                if (manager != null) {
                    manager.reformat()
                } else {
                    ApplicationUtils.invokeLater {
                        val fileEditorManager = FileEditorManager.getInstance(project)
                        val descriptor = OpenFileDescriptor(project, file)
                        val editor = ApplicationUtils.computableWriteAction {
                            fileEditorManager.openTextEditor(descriptor, false)
                        }
                        if (editor != null) {
                            service<EditorApplicationService>().forEditor(editor)?.reformat(closeAfter = true)
                        }
                    }
                }
            }
        }
    }

    /**
     * Reformat a file given its editor
     *
     * @param editor The editor
     */
    fun reformatFile(editor: Editor): Unit {
        service<EditorApplicationService>().forEditor(editor)?.reformat()
    }


    /**
     * Reformat a selection in a file given its editor
     *
     * @param editor The editor
     */
    fun reformatSelection(editor: Editor): Unit {
        service<EditorApplicationService>().forEditor(editor)?.reformatSelection()
    }

}